plugins/ssr/ssrequality.ml

```diff
@@ -336,17 +335,21 @@ let pirrel_rewrite ?(under=false) ?(map_redex=id_map_redex) pred rdx rdx_ty new_
   let sigma, p = (* The resulting goal *)
     Evarutil.new_evar env sigma (beta (EConstr.Vars.subst1 new_rdx pred)) in
   let pred = EConstr.mkNamedLambda (make_annot pattern_id Sorts.Relevant) rdx_ty pred in
-  let elim, gl =
-    let ((kn, i) as ind, _), unfolded_c_ty = pf_reduce_to_quantified_ind gl c_ty in
+  let elim, gl, sigma =
     let sort = elimination_sort_of_goal gl in
-    let elim, gl = pf_fresh_global (Indrec.lookup_eliminator env ind sort) gl in
-    if dir = R2L then elim, gl else (* taken from Coq's rewrite *)
-    let elim, _ = destConst elim in
-    let mp,l = Constant.repr2 (Constant.make1 (Constant.canonical elim)) in
-    let l' = Label.of_id (Nameops.add_suffix (Label.to_id l) "_r")  in
-    let c1' = Global.constant_of_delta_kn (Constant.canonical (Constant.make2 mp l')) in
-    mkConst c1', gl in
-  let elim = EConstr.of_constr elim in
+    match Equality.eq_elimination_ref (dir = L2R) sort with
```

Comment:
This match-expression does the following: If the elimination scheme has been registered then use it, otherwise use the "old" code to obtain the elimination scheme.

Reply:
It looks like the old code was wrong, so the new one is also wrong.

`gl` contains a `sigma` so one should thread, as a state, only `gl`, that is what is typically done in the rest of the code.

But here we have both `(sigma,c)` and `gl`, and given that we are using `c` to build `proof` (as well as `elim`) we should update `sigma` to obtain a new `(sigma,proof)`.

Imo the branch you copied, the one for `None` should also update sigma (and not `gl`) and use `Evd.fresh_global` instead of the lifter `pf_fresh_global`. Similarly, `c_ty` should be reduced using sigma via `Tacred.reduce_to_quantified_ind`.

Would you mind updating the code?

Reply:
Does it look right now?

Reply:
Great